test/test_bot.py

```diff
@@ -340,21 +341,53 @@ def url_handler(*args, **kwargs):
         return None
 
     sopel = bot.Sopel(tmpconfig, daemon=False)
-    # make sure we can always call it
-    sopel.unregister_url_callback(test_pattern)
 
     # now register a pattern, make sure it still work
     sopel.register_url_callback(test_pattern, url_handler)
     assert list(sopel.search_url_callbacks('https://www.example.com'))
 
     # unregister this pattern
-    sopel.unregister_url_callback(test_pattern)
+    sopel.unregister_url_callback(test_pattern, url_handler)
 
     # now it is not possible to find a callback for this pattern
     results = list(sopel.search_url_callbacks('https://www.example.com'))
     assert not results, 'Unregistered URL callback must not work anymore'
 
 
+def test_unregister_url_callback_no_memory(tmpconfig):
+    """Test unregister_url_callback behavior when bot.memory empty"""
+    test_pattern = r'https://(www\.)?example\.com'
+
+    def url_handler(*args, **kwargs):
+        return None
+
+    sopel = bot.Sopel(tmpconfig, daemon=False)
+    sopel.unregister_url_callback(test_pattern, url_handler)
+    # no exception implies success
+
+
+def test_unregister_url_callback_manual(tmpconfig):
+    """Test unregister_url_callback removes a specific callback that was added manually"""
```

Comment:
Here, I'd add a note saying that this can be removed once we deprecate the usage of callback added manually - which will happen soon enough.

Because otherwise, the next dev will wonder why we test the internal behavior of (un)register URL callback when it's clearly not our business.

Reply:
And by that I mean that this test is very good and add value for backward compatibility purpose, but let's not forget why it's good.

Reply:
Oh, agreed. Let's actually plan on moving the URL callback registry to a private member (`bot._url_callbacks`?) in 8.0, and stop supporting `bot.memory['url_callbacks']` at that time, along with removing this test.